### DISCUSSION TITLE
How to accomplish "fragments"?

### DISCUSSION THREAD

**Discussion by bravecrayon** [maizzle/framework#880] [Q&A] [answered]
I have a requirement to build separate "fragment" files that can be inserted into the final, rendered html email by the email-sending platform. These fragments would need to be built by Maizzle (css-inlined, etc) such that after inserting them into the email it would be one coherent file, css and all. Has anyone dealt with this before?

I've attempted to get this to work in various ways already, which I'll try to describe here.

First, I simply created these fragment files in the `src/templates` directory. This looked promising at first because these files were copied to the build directory. However, upon inspection, none of the css were inlined.

Second, I tried to make them a _real_ template that extends a very basic layout. The layout only had a `<block name="template"></block>` in it. Sadly, this yielded the same result as before. 

Then, I discovered that adding just `<style>{{{ page.css }}}</style>` to the fragment's layout file would trigger Maizzle into inlining the css. This works, but now there is a `<style>` tag with css tailored to each compiled fragment file. This could theoretically work, but is not ideal if multiple fragments are inserted into to the final, rendered html email. 

At this point I'm contemplating doing some scripting-fu. It works perfectly if I include each fragment as a Maizzle `<component>`. So, perhaps I could do that and also insert comments or custom tags as boundaries for a script to be able to find, cut out, and save to a separate filse in a post-build hook.

It's very possible i'm overthinking this or am not aware of a Maizzle feature or customization that I could leverage, so I thought I'd ask.

Thanks!

**Comment by cossssmin** [comment, marked as answer]
Had to do it once too, here's what I did.

Layout:

```html
<if condition="page.env === 'production'">
  <style>{{{ page.css }}}</style>
  <body>
    <block name="template"></block>
  </body>
</if>
<else>
  (normal layout with doctype etc., for developing locally and watching changes in the browser)
</else>
```

A template:

```html
<extends src="src/layouts/main.html">
  <block name="template">
    ....
  </block>
</extends>
```

Then in `config.production.js` remove the stuff you don't need. In my case I just removed the `<body>` tag but you could do the same for the `<style>` tag. The problem with removing `<style>` is you might have media queries or pseudo CSS that would be lost since it's not inlined.

```js
// config.production.js

module.exports = {
  // ...
  events: {
    afterTransformers(html) {
      return html.replace(/<\/?body.*?>/g, '')
    }
  },
}
```

**Reply by cossssmin** [reply]
You need the `<body>` tag for `removeUnusedCSS` to work properly, it won't remove unused classes after inlining otherwise.

**Reply by bravecrayon** [reply]
@cossssmin Thanks so much for that quick response and those ideas. After reading this, I decided to just try using the really simple template `<style>{{{ page.css }}}</style><block name="template"></block>` for each of the fragments and just leaving in the styles that are generated. I was concerned at first that including those style tags, possibly multiple of them and between the rows of a table, might be problematic for some clients, but it doesn't not appear so! Inefficient maybe, but I must draw a line somewhere to get the work done.

**Reply by cossssmin** [reply]
Hmm now that you mention it again, there will be an issue with `<style>` blocks inside `<body>` - Gmail doesn't support them:

https://www.caniemail.com/features/html-style/

In Maizzle that could be fixed with a custom PostHTML transformer that would take all `<style>` tags and move them to the `<head>`, but I understand you're not piecing the entire email together in Maizzle, are you?

**Reply by bravecrayon** [reply]
@cossssmin At first I didn't see any issues with including the `<style>` blocks but, yep, they eventually became apparent in the mobile  views for me. Thanks for the heads-up.

Since I didn't know how to do a custom PostHTML transformer, I decided to try a combination of including the fragments as components in the main email body, adding custom tags around each fragment's html to differentiate them, and use `afterTransformers` to extract each bit of fragment html into a separate file in `build_production`. Because the fragments are included in the template, all of the necessary styles end up the main `<style>` block for the email, which I didn't mind.

If anyone else stumbles upon this thread, this is essentially what I did.

I made each component output with a custom tag around the fragment's html: `<fragment name="{{name}}">...</fragment>`

The `name` is set by from the component reference: `<component src="/src/components/fragment.html" name="01 -- Some name for fragement file"></component>`. The name becomes the filename.

When built it I get files like this: `build_production/fragments/01 -- Some name for fragement file.html`

Here's the `afterTransformers`:

```js
events: {
    async afterTransformers(html, c) {
      
      const r = /<fragment name="(?<name>.*?)">(?<frag>.*?)<\/fragment>/gsm;
      for (m of html.matchAll(r)) {
        const destination = `${c.build.templates.destination.path}/fragments/${m.groups.name.replace("\n", ' ')}.html`
        await fs.outputFile(destination, m.groups.frag)
      }

      html = html.replaceAll(r, '')
      return html;
    }
  }
```

Elegant or not it worked for me.
